### PR TITLE
rbd: merge diff files

### PR DESCRIPTION
**Pull request by dragonylffly**
RBD supplies the 'export-diff' to export a diff file between two clones.
This is inconvinient for periodical backup, since a bunch of diff files 
wastes the disk space by keeping multiple copies of the data on the 
same location, and complicates and slows down the recovery process.
In this case, only one diff file is expected. This motivates us to develop 
a 'merge-diff' tool. The semantic is straightforward, it merges two diff 
files into one single diff  file. Multiple diff files could be easily merged 
by further using some simple shell tricks.

### CODE REVIEW

**Comment by liewegas** [comment]
i think the code should be restructured a bit to not assume the header tag order puts 's' last.  probably need to read the next tag to tell when the header is done which will be a bit awkward (but hopefully not too bad?)

**Comment by wwformat** [comment]
> i think the code should be restructured a bit to not assume the header tag order puts 's' last. probably need to read the next tag to tell when the header is done which will be a bit awkward (but hopefully not too bad?)

OK. 
I have already fix these, and pushed the new code to here.

File: src/rbd.cc [hunk 1]
```diff
@@ -1708,6 +1712,352 @@ static int do_import_diff(librbd::Image &image, const char *path)
...
+      ::decode(*size, p);
+    } else {
+      break;
+    }
+  }
+
+  return 0;
+}
+
+static int parse_diff_body(int fd, __u8 *tag, uint64_t *offset, uint64_t *length)
```
Comment (liewegas) [inline src/rbd.cc]: this makes sense: teh caller can either provide the tag it already read or we will read it ourselves.

File: src/rbd.cc [hunk 2]
```diff
@@ -1708,6 +1712,352 @@ static int do_import_diff(librbd::Image &image, const char *path)
...
+  //We just handle the case like 'banner, [ftag], [ttag], stag, [wztag]*,etag',
+  // and the (offset,length) in wztag must be ascending order.
+
+  r = parse_diff_header(fd, &f_from, &f_to, &f_size);
+  if (r < 0)
+    goto done;
+
+  r = parse_diff_header(sd, &s_from, &s_to, &s_size);
+  if (r < 0)
+    goto done;
```
Comment (liewegas) [inline src/rbd.cc]: ...but, parse_diff_header throws out the final tag it reads.  shouldn't we pass in &f_tag and &s_tag?

File: src/rbd.cc [hunk 3]
```diff
@@ -1708,6 +1712,352 @@ static int do_import_diff(librbd::Image &image, const char *path)
...
+      r = safe_read_exact(fd, buf, 8);
+      if (r < 0)
+        return r;
+
+      bufferlist bl;
+      bl.append(buf, 8);
+      bufferlist::iterator p = bl.begin();
+      ::decode(*size, p);
+    } else {
+      break;
```
Comment (liewegas) [inline src/rbd.cc]: as far as I can tell the last tag we read is thrown out... how does the caller know what the next chunk is?

Comment (wwformat) [inline src/rbd.cc]: > as far as I can tell the last tag we read is thrown out... how does the caller know what the next chunk is?

Sorry, I make a mistake in using git to modify commits. ^_^
In my desigin, parse_diff_header will parse the first tag with data block,
and the parse_diff_body will try to reuse the tag by the result of parse_diff_header.

**Comment by dragonylffly** [comment]
The commits have been updated, please review

File: src/rbd.cc [hunk 4]
```diff
@@ -1708,6 +1712,351 @@ static int do_import_diff(librbd::Image &image, const char *path)
...
+  if (*tag == 'e') {
+    offset = 0;
+    length = 0;
+    return 0;
+  }
+
+  if (*tag != 'w' && *tag != 'z')
+    return -ENOTSUP;
+
+  char buf[16];
```
Comment (liewegas) [inline src/rbd.cc]: in the 'z' case we shouldn't read anything.. it's all zeros

Comment (wwformat) [inline src/rbd.cc]: > in the 'z' case we shouldn't read anything.. it's all zeros

from the routine in 'export_diff_cb'

Yes, after read the 'z' tag, the code just only read 16 bits with offset and length.

Comment (liewegas) [inline src/rbd.cc]: nevermind , i see now

**Comment by liewegas** [comment]
looks good to me!  Reviewed-by:

File: src/rbd.cc (hunk 4 above)
Comment (wwformat) [inline src/rbd.cc]: > in the 'z' case we shouldn't read anything.. it's all zeros

By the routine 'export_diff_cb', the data after the 'z' tag is 16 bits value with offset and length.

File: src/rbd.cc [hunk 5]
```diff
@@ -2385,7 +2741,8 @@ if (!set_conf_param(v, p1, p2, p3)) { \
       opt_cmd != OPT_IMPORT &&
       opt_cmd != OPT_IMPORT_DIFF &&
       opt_cmd != OPT_UNMAP &&
-      opt_cmd != OPT_SHOWMAPPED && !imgname) {
+      opt_cmd != OPT_SHOWMAPPED &&
+      opt_cmd != OPT_MERGE_DIFF && !imgname) {
```
Comment (jdurgin) [inline src/rbd.cc]: it doesn't look like image name is required

File: src/test/cli/rbd/help.t [hunk 6]
```diff
@@ -25,6 +25,10 @@
     export-diff <image-name> [--from-snap <snap-name>] <path>
                                                 export an incremental diff to
                                                 path, or "-" for stdout
+    merge-diff <diff1> <diff2> <path>           merge <diff1> and <diff2> into
```
Comment (jdurgin) [inline src/test/cli/rbd/help.t]: could you add this to doc/man/8/rbd.rst (the man page source) too?

File: src/rbd.cc (hunk 5 above)
Comment (wwformat) [inline src/rbd.cc]: > it doesn't look like image name is required

Yes, so when the opt_cmd is OPT_MERGE_DIFF, this code will ignore to check imgname.

Comment (jdurgin) [inline src/rbd.cc]: right, I misread that

File: src/test/cli/rbd/help.t (hunk 6 above)
Comment (wwformat) [inline src/test/cli/rbd/help.t]: > could you add this to doc/man/8/rbd.rst (the man page source) too?

ok, I already pushed a commit for this.
please review it.

**Comment by wwformat** [comment]
> FAIL: the output of run-make-check.sh on centos-centos7 for 958ab8c is http://paste2.org/Wh68kLfz
> 
> :octocat: Sent from GH.

the problem seems come from master branch.

**Comment by dachary** [comment]
@wwformat yes, you are correct http://tracker.ceph.com/issues/10547

**Comment by jdurgin** [comment]
Thanks, looks great!

**Pull request merged**